### PR TITLE
Feat/chain-id

### PR DESCRIPTION
**Pull request by jcnelson**
This PR address #2982 by adding `chain-id` as a global variable in `Clarity2`, which resolves to `StacksChainState::chain_id`.

### CODE REVIEW

File: clarity/src/vm/tests/simple_apply_eval.rs [hunk 1]
```diff
@@ -1424,3 +1426,46 @@ fn test_is_mainnet() {
...
+    tests
+        .iter()
+        .zip(mainnet_expectations.iter())
+        .for_each(|(program, expectation)| {
+            assert_eq!(
+                expectation.clone(),
+                execute_with_parameters(
+                    program,
+                    ClarityVersion::Clarity2,
+                    StacksEpochId::Epoch20,
```
Comment (gregorycoppola) [inline clarity/src/vm/tests/simple_apply_eval.rs:1445]: why not Epoch21? same as below.

**Review by gregorycoppola — APPROVED** [review]

File: clarity/src/vm/tests/simple_apply_eval.rs (hunk 1 above)
Comment (jcnelson) [reply clarity/src/vm/tests/simple_apply_eval.rs:1445]: Ah, that's an oversight on my end.  Will fix.

File: clarity/src/vm/contexts.rs [hunk 2]
```diff
@@ -204,6 +206,8 @@ pub struct GlobalContext<'a> {
     pub coverage_reporting: Option<CoverageReporter>,
     /// This is the epoch of the the block that this transaction is executing within.
     epoch_id: StacksEpochId,
+    /// This is the chain ID of the transaction
+    pub chain_id: u32,
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply clarity/src/vm/contexts.rs:210]: No -- `mainnet` and `chain_id` are orthogonal concepts.  The fact that we have a separate `chain_id` for our testnet is a design accident -- the transaction wire format has separate fields for mainnet/testnet and chain IDs.

File: src/clarity_vm/tests/analysis_costs.rs [hunk 3]
```diff
@@ -43,10 +43,16 @@ use crate::clarity_vm::database::marf::MarfedKV;
...
+use stacks_common::consts::{CHAIN_ID_MAINNET, CHAIN_ID_TESTNET};
 
 pub fn test_tracked_costs(prog: &str, use_mainnet: bool, epoch: StacksEpochId) -> ExecutionCost {
     let marf = MarfedKV::temporary();
-    let mut clarity_instance = ClarityInstance::new(use_mainnet, marf);
+    let chain_id = if use_mainnet {
+        CHAIN_ID_MAINNET
+    } else {
+        CHAIN_ID_TESTNET
+    };
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply src/clarity_vm/tests/analysis_costs.rs:54]: Moreover, the eight instances of this code path only occur in tests.  There should be no place in prod where we derive the chain ID from the `mainnet` boolean, or vice versa.

[the remark this answers is not in the document]
Comment (jcnelson) [reply src/clarity_vm/tests/analysis_costs.rs:54]: If you'd like, I can create `#[test]`-compiled function that has a very long and explicit docstring that warns the reader of this.

File: clarity/src/vm/contexts.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (obycode) [reply clarity/src/vm/contexts.rs:210]: Okay, thanks for that clarification. Makes sense.

File: src/clarity_vm/tests/analysis_costs.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jcnelson) [reply src/clarity_vm/tests/analysis_costs.rs]: Added

[the remark this answers is not in the document]
Comment (obycode) [reply src/clarity_vm/tests/analysis_costs.rs]: Thanks. I think that's helpful.

File: clarity/src/vm/contexts.rs [hunk 4]
```diff
@@ -51,6 +51,10 @@ use crate::vm::version::ClarityVersion;
 
 use crate::vm::coverage::CoverageReporter;
 
+use crate::vm::tests::test_only_mainnet_to_chain_id;
```
Comment (obycode) [inline clarity/src/vm/contexts.rs]: Should this only be imported in test?

File: clarity/src/vm/mod.rs [hunk 5]
```diff
@@ -428,8 +430,18 @@ pub fn execute_with_parameters(
     let mut contract_context = ContractContext::new(contract_id.clone(), clarity_version);
     let mut marf = MemoryBackingStore::new();
     let conn = marf.as_clarity_db();
-    let mut global_context =
-        GlobalContext::new(use_mainnet, conn, LimitedCostTracker::new_free(), epoch);
+    let chain_id = if use_mainnet {
+        CHAIN_ID_MAINNET
+    } else {
+        CHAIN_ID_TESTNET
+    };
```
Comment (obycode) [inline clarity/src/vm/mod.rs:437]: Can use `test_only_mainnet_to_chain_id` here (and remove const imports).

File: src/clarity_vm/tests/analysis_costs.rs [hunk 6]
```diff
@@ -42,11 +42,14 @@ use crate::clarity_vm::clarity::ClarityConnection;
 use crate::clarity_vm::database::marf::MarfedKV;
 use crate::types::chainstate::{BlockHeaderHash, StacksBlockId};
 use crate::types::StacksEpochId;
+use clarity::vm::tests::test_only_mainnet_to_chain_id;
 use clarity::vm::ClarityVersion;
+use stacks_common::consts::{CHAIN_ID_MAINNET, CHAIN_ID_TESTNET};
```
Comment (obycode) [inline src/clarity_vm/tests/analysis_costs.rs:47]: These are unused now

File: src/clarity_vm/tests/events.rs [hunk 7]
```diff
@@ -27,11 +27,14 @@ use crate::vm::tests::execute;
 use crate::vm::tests::{TEST_BURN_STATE_DB, TEST_HEADER_DB};
 use crate::vm::types::{AssetIdentifier, BuffData, QualifiedContractIdentifier, Value};
 
+use stacks_common::consts::{CHAIN_ID_MAINNET, CHAIN_ID_TESTNET};
```
Comment (obycode) [inline src/clarity_vm/tests/events.rs:30]: These are unused now.

**Review by obycode — COMMENTED** [review]
Thanks for that change with `test_only_mainnet_to_chain_id`.

File: clarity/src/vm/contexts.rs (hunk 4 above)
Comment (obycode) [reply clarity/src/vm/contexts.rs]: Fixed in recent commit.

File: clarity/src/vm/mod.rs (hunk 5 above)
Comment (jcnelson) [reply clarity/src/vm/mod.rs]: Fixed

File: src/clarity_vm/tests/analysis_costs.rs (hunk 6 above)
Comment (jcnelson) [reply src/clarity_vm/tests/analysis_costs.rs]: Fixed

File: src/clarity_vm/tests/events.rs (hunk 7 above)
Comment (jcnelson) [reply src/clarity_vm/tests/events.rs]: Fixed

**Review by obycode — APPROVED** [review]
LGTM!

**Pull request merged**